ebpfcore/ebpf_drv.c

```diff
@@ -406,4 +239,202 @@ _ebpf_driver_query_volume_information(_In_ WDFDEVICE device, _Inout_ IRP* irp)
+
+    WDF_DRIVER_CONFIG_INIT(&driver_configuration, WDF_NO_EVENT_CALLBACK);
+    driver_configuration.DriverInitFlags |= WdfDriverInitNonPnpDriver;
+    driver_configuration.EvtDriverUnload = _ebpf_driver_unload;
+    status =
+        WdfDriverCreate(driver_object, registry_path, WDF_NO_OBJECT_ATTRIBUTES, &driver_configuration, driver_handle);
+    if (!NT_SUCCESS(status)) {
+        EBPF_LOG_NTSTATUS_API_FAILURE(EBPF_TRACELOG_KEYWORD_ERROR, "WdfDriverCreate", status);
+        goto Exit;
+    }
+
+    status = _ebpf_driver_initialize_device(*driver_handle, device);
+    if (!NT_SUCCESS(status)) {
+        EBPF_LOG_MESSAGE_NTSTATUS(
+            EBPF_TRACELOG_LEVEL_CRITICAL, EBPF_TRACELOG_KEYWORD_ERROR, (char*)"_ebpf_driver_initialize_device", status);
+        goto Exit;
+    }
+
+    device_create_flag = TRUE;
+
+    // create default queue
+    WDF_IO_QUEUE_CONFIG_INIT_DEFAULT_QUEUE(&io_queue_configuration, WdfIoQueueDispatchParallel);
+    io_queue_configuration.EvtIoDeviceControl = _ebpf_driver_io_device_control;
+    status = WdfIoQueueCreate(*device, &io_queue_configuration, WDF_NO_OBJECT_ATTRIBUTES, WDF_NO_HANDLE);
+    if (!NT_SUCCESS(status)) {
+        EBPF_LOG_NTSTATUS_API_FAILURE(EBPF_TRACELOG_KEYWORD_ERROR, "WdfIoQueueCreate", status);
+        goto Exit;
+    }
+
+    status = ebpf_result_to_ntstatus(ebpf_core_initiate());
+    if (!NT_SUCCESS(status)) {
+        EBPF_LOG_NTSTATUS_API_FAILURE(EBPF_TRACELOG_KEYWORD_ERROR, "ebpf_core_initiate", status);
+        goto Exit;
+    }
+
+    WdfControlFinishInitializing(*device);
+
+Exit:
+    if (!NT_SUCCESS(status)) {
+        if (device_create_flag && device != NULL) {
+
+            // Release the reference on the newly created object, since we couldn't initialize it.
+            WdfObjectDelete(*device);
+        }
+    }
+    return status;
+}
+
+NTSTATUS
+DriverEntry(_In_ DRIVER_OBJECT* driver_object, _In_ UNICODE_STRING* registry_path)
+{
+    NTSTATUS status;
+    WDFDRIVER driver_handle;
+    WDFDEVICE device;
+
+    status = ebpf_trace_initiate();
+    if (!NT_SUCCESS(status)) {
+
+        // Fail silently as there is no other mechanism to indicate this failure. Note that in this case, the
+        // EBPF_LOG_EXIT() call at the end will not log anything either.
+        goto Exit;
+    }
+
+    EBPF_LOG_ENTRY();
+
+    // Request NX Non-Paged Pool when available
+    ExInitializeDriverRuntime(DrvRtPoolNxOptIn);
+    status = _ebpf_driver_initialize_objects(driver_object, registry_path, &driver_handle, &device);
+    if (!NT_SUCCESS(status)) {
+        EBPF_LOG_MESSAGE_NTSTATUS(
+            EBPF_TRACELOG_LEVEL_CRITICAL,
+            EBPF_TRACELOG_KEYWORD_ERROR,
+            (char*)"_ebpf_driver_initialize_objects failed",
+            status);
+        goto Exit;
+    }
+
+    _ebpf_driver_device_object = WdfDeviceWdmGetDeviceObject(device);
+
+Exit:
+    EBPF_LOG_EXIT();
+    return status;
+}
+
+// Needed by the statically kernel lib(s)
+DEVICE_OBJECT*
```

Comment:
Can we annotate the return value with either `_Ret_notnull_` or `_Ret_maybenull_` depending on which is the case?